File: python3/Linked Lists/flatten_multi_level_list.py
from ds import MultiLevelListNode

"""
Definition of MultiLevelListNode:
class MultiLevelListNode:
    def __init__(self, val, next, child):
        self.val = val
        self.next = next
        self.child = child
"""


def flatten_multi_level_list(head: MultiLevelListNode) -> MultiLevelListNode:
    if not head:
        return None
    tail = head
    # Find the tail of the linked list at the first level.
    while tail.next:
        tail = tail.next
    curr = head
    # Process each node at the current level. If a node has a child linked list,
    # append it to the tail and then update the tail to the end of the extended 
    # linked list. Continue until all nodes at the current level are processed.
    while curr:
        if curr.child:
            tail.next = curr.child
            # Disconnect the child linked list from the current node.
            curr.child = None
            while tail.next:
                tail = tail.next
        curr = curr.next
    return head

File: typescript/Linked Lists/flatten_multi_level_list.ts
import { MultiLevelListNode } from "./ds";

/* Definition of MultiLevelListNode: 
class MultiLevelListNode {
    val: number; 
    next: MultiLevelListNode | null; 
    child: MultiLevelListNode | null;
    constructor(val: number){
        this.val = val;
        this.next = null;
        this.child = null;
    }
}*/


function flattenMultiLevelList(head: MultiLevelListNode | null): MultiLevelListNode | null {
    if (!head) 
        return head;
    let tail: MultiLevelListNode | null = head;
    // Find the tail of the linked list at the first level.
    while (tail.next !== null)
        tail = tail.next;
    let curr: MultiLevelListNode | null = head;
    // Process each node at the current level. If a node has a child linked list,
    // append it to the tail and then update the tail to the end of the extended 
    // linked list. Continue until all nodes at the current level are processed.
    while (curr !== null){
        if (curr.child !== null){
            tail.next = curr.child;
            // Disconnect the child linked list from the current node.
            curr.child = null;
            while (tail.next !== null)
                tail = tail.next;
        }
        curr = curr.next;
    }
    return head;
}